### ISSUE TITLE
The way to benchmark multithreaded actions.

### ISSUE THREAD

**Issue by kayanme** [PerfDotNet/BenchmarkDotNet#147]
Consider some object, which should hold a lot of concurrent operations at the same time.
There is the need to benchmark one operation from one thread (still having others working) to verify  productivity during parallel load.
The problem is how to correctly hold that background threads still measuring one needed.

I see some ways, but none of them are completely possible now.
1. An ability to run benchmark methods in parallel, so every thread runs its own measuring still providing the needed concurrency.
2. Establish background in setup method. Unfortunatly, there is a need of "unsetup" method, where threads should be stopped.

**Comment by AndreyAkinshin** [comment]
@kayanme, this is a very great idea. We definitely should add this feature. Unfortunately, it is hard to implement right way. There are a lot of multi-threading side effects like false sharing which can easily spoil a benchmark. We will add it sometimes, but not in the nearest future.

**Comment by jmbryan22** [comment]
#https://github.com/danielpalme/IocPerformance/issues/77

**Comment by jmbryan22** [comment]
danielpalme/IocPerformance#77

**Comment by roji** [comment]
Just a big +1 to adding support for properly benchmarking in a concurrent environment - it's super important to be able to see how code behaves under concurrent load (lock contention, CPU caches...). Some knobs would be necessary - number of threads etc.

**Comment by spati2** [comment]
+1

**Comment by nxrighthere** [comment]
This should be one of the top priorities, please.

**Comment by lydonchandra** [comment]
+1

**Comment by osjoberg** [comment]
+1

**Comment by theEyeD** [comment]
While not an exact replacement, with the [deprecation of the Visual Studio load tests](https://github.com/MicrosoftDocs/visualstudio-docs/issues/2032), it would be great to be able to migrate some the loads tests to benchmarks and get some similar data out of it (customize # of 'users'/threads, time per run, etc.).

**Comment by kirsan31** [comment]
Any news/thoughts on this?

**Comment by ashishnegi** [comment]
I have simpler scenario where i want to benchmark operations like `Add` on `ConcurrentDictionary` like collections, parameterizing the number of threads. i.e. how does this collection's a particular operation scale with the concurrent requests ? For inspiration, google's [benchmark](https://github.com/google/benchmark) library support this.

**Comment by lupengamzn** [comment]
+1

**Comment by mikehung** [comment]
+1

**Comment by timcassell** [comment]
I wrote a helper to make it simpler and more efficient to benchmark multi-threaded actions then manually setting up threads. https://github.com/timcassell/ProtoBenchmarkHelpers This can be used until BDN adds multi-threaded benchmarking officially.

**Comment by Stabzs** [comment]
@timcassell does your helper work for async actions? The generation of async state machines seems to wreak havoc on all BDN benchmarking and while your helper is great for running multiple threads in parallel, it doesn't seem to be sufficient for async methods. Either way, thanks for your work, it is fantastic!

**Comment by timcassell** [comment]
@Stabzs I did not have async in mind when I wrote that. I'm not sure adding async support would be of much use, though. If your async actions hop to different threads while they're running, it all goes out the window, and you might as well just use `Task.Run` and `Task.WhenAll`. Do you have a specific use-case in mind?

**Comment by Stabzs** [comment]
@timcassell  thanks for the quick response!

There's a handful of scenarios that I'd like to benchmark where I need to run multiple concurrent tasks to simulate contention across threads, such as the memory cost of asynchronously waiting on a potentially contended `SemaphoreSlim` instance. Some optimizations (pooling, ValueTask, etc etc), can have significant allocation reductions. However, the overhead from arrays of tasks and `Task.Run` skew the allocations of the asynchronous workers themselves, making it difficult to measure performance.

In addition, both allocations and elapsed time tend to be very sporadic when benchmarking async methods.

Overall, it doesn't seem that there is a great way to benchmark async methods. I've used your helper before, but in hindsight, it probably is the wrong approach.

**Comment by timcassell** [comment]
> In addition, both allocations and elapsed time tend to be very sporadic when benchmarking async methods.

Well that's not surprising if they're hopping to different threads.

> Some optimizations (pooling, ValueTask, etc etc), can have significant allocation reductions. However, the overhead from arrays of tasks and `Task.Run` skew the allocations of the asynchronous workers themselves, making it difficult to measure performance.

That's a good point. I have an idea how to support this. I'll open a new issue on that repo for this.

**Comment by timcassell** [comment]
@Stabzs In the meantime, you could try synchronously waiting on the tasks inside the actions. `task.GetAwaiter().GetResult()`. That's how BDN currently consumes tasks anyway.

**Comment by Stabzs** [comment]
@timcassell  I appreciate the suggestion! I'll give that a try and see if it changes the results.

**Comment by timcassell** [comment]
No-one has suggested what the API for this should look like yet, so here's my idea.

```cs
public class Benchmarks
{
    [ParamsAllValues]
    public bool MyFlag { get; set; }

    [Benchmark]
    public void SingleThread() { }

    [ThreadedBenchmark(MaxConcurrency = 2)] // Default MaxConcurrency is Environment.ProcessorCount
    public class MultiThread : Benchmarks // Inherit from the parent type to have access to the same params as regular benchmarks
    {
        [ThreadInvoke] // Invoke once on a thread
        public void Action() { }

        [ThreadInvoke(3)] // Invoke 3 times, each on a different thread
        public int Func() => 42;
        
        // Support the same return types as regular benchmark methods
        [ThreadInvoke]
        public async Task AsyncAction() { }
    }
}
```

That would produce a table like this:

|       Method | MyFlag |    Mean |
|------------- |------- |--------:|
| SingleThread |  False | 0.00 ns |
|  MultiThread |  False | 0.00 ns |
| SingleThread |   True | 0.00 ns |
|  MultiThread |   True | 0.00 ns |

This encapsulates the threaded actions within a child class (even though the result table labels it as a `Method`, I think it's quite readable).

**Comment by CDboyOne** [comment]
+1 We also need multi-threading benchmark very much

**Comment by triynko** [comment]
Need a simple way to multi-thread.

ChatGPT thought this might already be supported with syntax like:
// Run this method concurrently using 4 threads
    [Benchmark]
    [Threading(4)]
    public void NonThreadedMethod()

But it was just being creative.

We need a basic way to have a method run, not just in a loop on one thread, but called from multiple threads.  The real reason we want to do this isn't necessarily get accurate performance benchmarks, but rather just to use Benchmark.NET as a 'runner' for performance profiling. We use the 'InProcess' option to have it run the benchmarks within the same process as Visual Studio's profiler, so profiling becomes as simple as marking a method with 'Benchmark' attribute, then running the VS built-in profiler. We'd just like a simple way to ensure the benchmarking happens on multiple threads simultaneously. This will help us test things like object pooling behaviors.
